File: HW5/boardArray2D.cpp
/*File boardArray2D.cpp
 *This is the implementation file for class BoardArray2D*/

#include "boardArray2D.h"

namespace EsraNameSpace{

	//No parameter constructor
	BoardArray2D::BoardArray2D(): row(9),column(9)
	{
		arr2D = new int*[9];
		for(int i=0; i<row;i++){
			arr2D[i]=new int[9];
		}
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				arr2D[i][j]=-1;
			}
		}
		setNumOfBoard();
		setLastMove('S');
	}

	//Constructor takes row and column
	BoardArray2D::BoardArray2D(int nrow,int ncolumn): row(nrow),column(ncolumn)
	{
		arr2D = new int*[row];
		for(int i=0; i<row;i++){
			arr2D[i]=new int[column];
		}
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				arr2D[i][j]=-1;
			}
		}
		setNumOfBoard();
		setLastMove('S');
	}
	//Constructor takes size
	BoardArray2D::BoardArray2D(int size): row(size),column(size)
	{
		arr2D = new int*[size];
		for(int i=0; i<size;i++){
			arr2D[i]=new int[size];
		}
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				arr2D[i][j]=-1;
			}
		}
		setNumOfBoard();
		setLastMove('S');
	}

	//Destructor
	BoardArray2D::~BoardArray2D(){
		for (int i = 0; i < row; i++)
  			delete [] arr2D[i]; 
		delete [] arr2D;
	}
	//Copy constructor
	BoardArray2D::BoardArray2D(const BoardArray2D& object){
		row=object.getRow();
		column=object.getColumn();
		numOfMovesPerformed=object.numberOfMoves();
		numOfBoardsCreated=object.numberOfBoards();
		lastMovePerf=object.lastMove();
		arr2D = new int*[row];
		for(int i=0; i<row;i++){
			arr2D[i]=new int[column];
		}
		for(int i=0; i<row;i++){
			for(int j=0; j<column;j++){
				arr2D[i][j]=object.getIndex(i,j);
			}
		}

	}
	//Assignment operator
	BoardArray2D& BoardArray2D::operator =(const BoardArray2D& rightSide){
		//Error check for control objects are same size
		if (row= rightSide.getRow() && column==rightSide.getColumn()){
			for(int i=0; i<row;i++){
				for(int j=0; j<row;j++){
					arr2D[i][j]=rightSide.getIndex(i,j);
				}
			}
		}
		else
			cout<<"Objects are not same size!";
	}
	//Prints the board by sending it to cout
	void BoardArray2D::print(){

		for (int m = 0; m < row; m++){
	        for (int k = 0; k < column; k++) {
	          if(arr2D[m][k]!=100)
	            cout << arr2D[m][k] << " ";
	          else if(arr2D[m][k]==100)
	            cout<<"bb ";
	        }
	        cout<<endl;
	  }
	  cout<<endl;
	}

	//Reads file and determines new row and column informations
	void BoardArray2D::readFromFile(ifstream &input,string filename){
		
		input.open(filename.c_str());
		char readen[100];
	  	char *point;       //will used for strtok
	  	int i=0,j=0;

		if(input.fail()){
		    cerr<<"Error opening file!";
		    exit(1);
		}
		while(!input.eof()){

		    input.getline(readen,100);
		    point=strtok(readen," ");

		    while(point!=NULL){

		      point=strtok(NULL," ");
		      j++;
		    }
		    i++;
		}
		//determining row and column numbers according to file
		setRow(i);
		setColumn(j/i);
		input.close();
	}	

	//Reads file and assigns the values from file to board			
	void BoardArray2D::readAndAssign(ifstream &input,string filename){

		delete [] arr2D;

		arr2D = new int*[getRow()];
		for(int i=0; i<getRow();i++){
			arr2D[i]=new int[getColumn()];
		}
		input.open(filename.c_str());
		std::string readen;
		int i=0,j=0;
		int number=0;
		if(input.fail()){
		cerr<<"Error opening file!";
		exit(1);
		}
		
		while(!input.eof()){
			for(i=0;i<row;++i){
			  for(j=0;j<column;++j){
			    input>>readen;
			    if(readen!="bb"){
			      //Readen is type of string so i convert it to integer
			      number=atoi(readen.c_str());
			     
			      arr2D[i][j]=number;
			    }
			    if(readen=="bb"){
			    	
			      arr2D[i][j]=100;
			    } 
			  }
			}
  		}

	}

	//Writes the board configuration to given file
	void BoardArray2D::writeToFile(ofstream &output,string &filename){
		
		output.open(filename.c_str()); 
		if(output.fail()){
			cerr<<"Error opening file!";
			exit(1);
		}
		if(output.is_open()){

			for(int i=0; i<row ;i++){
				for(int j=0; j<column; j++){
					if(arr2D[i][j]!=100 && arr2D[i][j]>9){
						output<<arr2D[i][j];
						if(j!=column-1)
							output<<" ";
					}

					else if(arr2D[i][j]<10){
						output<<"0"<<arr2D[i][j];
						if(j!=column-1)
							output<<" ";
					}
					else if(arr2D[i][j] == 100){
						output<<"bb";
						if(j!=column-1)
							output<<" ";
					}
				}
				if(i!=row-1)
					output<<"\n";
			}
		}
		else 
		cerr<<"Error opening file!\n";
		output.close();
	}	

	//Resets board to the solution	
	void BoardArray2D::reset(){
		
		int num=1;
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				if(i==row-1 && j==column-1 && arr2D[i][j]!=0){

					arr2D[i][j]=100;
				}
				else if(arr2D[i][j]!=0){

					arr2D[i][j]=num;
					num++;
				} 
			}
		}
	}

	//Sets the size and resets the board
	void BoardArray2D::setSize(int rowS,int columnS){
		
		setRow(rowS);
		setColumn(columnS);
		delete [] arr2D;

		arr2D = new int*[getRow()];
		for(int i=0; i<getRow();i++){
			arr2D[i]=new int[getColumn()];
		}
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				arr2D[i][j]=-1;
			}
		}
		reset();
	}

	//Moves the empty cell to the given direction
	int BoardArray2D::move(char direction){
		int emptCellX, emptCellY;
		int num;

		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++){
				if(arr2D[i][j]==100){
					emptCellX=i;
					emptCellY=j;
				}
			}
		}
		if (direction=='U'){
			
			if(emptCellX==0 || arr2D[emptCellX-1][emptCellY]==0){
				return -1;
			}
			else if(emptCellX!=0 && arr2D[emptCellX-1][emptCellY]!=0){
				num=arr2D[emptCellX-1][emptCellY];
				arr2D[emptCellX-1][emptCellY]=100;
				arr2D[emptCellX][emptCellY]=num;
				setLastMove('U');
				setNumOfMove();
			}
			return 0;  
		}
		else if(direction== 'D'){

			if(emptCellX==row-1 || arr2D[emptCellX+1][emptCellY]==0){
				return -1; 
			}
			else if(emptCellX!=row-1 && arr2D[emptCellX+1][emptCellY]!=0){
				num=arr2D[emptCellX+1][emptCellY];
				arr2D[emptCellX+1][emptCellY]=100;
				arr2D[emptCellX][emptCellY]=num;
				setLastMove('D');
				setNumOfMove();
			} 
			return 0;
		}
		else if(direction=='R'){

			if(emptCellY==column-1 || arr2D[emptCellX][emptCellY+1]==0){
				return -1;
			}

			else if(emptCellY!=column-1 && arr2D[emptCellX][emptCellY+1]!=0){
				num=arr2D[emptCellX][emptCellY+1];
				arr2D[emptCellX][emptCellY+1]=100;
				arr2D[emptCellX][emptCellY]=num;
				setLastMove('R');
				setNumOfMove();
			} 
			return 0;
		}
		else if(direction=='L'){

			if(emptCellY==0 || arr2D[emptCellX][emptCellY-1]==0){ 
				return -1;
			}

			else if(emptCellY!=0 && arr2D[emptCellX][emptCellY-1]!=0){
				num=arr2D[emptCellX][emptCellY-1];
				arr2D[emptCellX][emptCellY-1]=100;
				arr2D[emptCellX][emptCellY]=num;
				setLastMove('L');
				setNumOfMove();
			} 
			return 0;
		}
	}

	//Shuffles the board
	int BoardArray2D::shuffle(){
		/*First reset the board to the solution*/
		reset();
		
		int randomMove;
		int control=0;
		int i=0;
		/*This loop written for shuffle the board making ramdom moves size*size.
		But if there is an impossible move, i call random func. again to make sure
		the board is shuffled properly */
		while( i<row*column){

		randomMove = (rand() % 4)+1 ;
		if(randomMove==1){
			control=move('U');
		} 
		if(randomMove==2){
			control=move('D');
		} 
		if(randomMove==3){
			control=move('L');
		} 
		if(randomMove==4){
			control=move('R');
		} 
		/*Control is the return value for U,D,R,L functions
		If control is -1 this means there is no room for that choice
		If control is 0 this means cell is moved and size*size number increases*/
		if(control==0)
			i++;
		}
		//cout<<"Shuffled version:\n";
		//print();
		return 0;
	}

	//Returns true if board is solved
	bool BoardArray2D::isSolved(){
		BoardArray2D controlObje(row,column);
		controlObje.reset();

		int count=0;
		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++){
				if(arr2D[i][j]==controlObje.getIndex(i,j)){
					count++;
				}
				else if(arr2D[i][j]!=controlObje.getIndex(i,j)){
					cout<<"Problem is not solved.\n";
					return false;
				}
			}
		}
		if(count==row*column) {
			cout<<"Problem is solved!\n";
			return true;
		}
	}

	//Returns true if two boards are equal
	bool BoardArray2D::operator ==(const BoardArray2D& rightSide)  {

		int count=0;

		for(auto i=0; i<row;i++){
			for(auto j=0; j<column;j++){
				if(arr2D[i][j]==rightSide.getIndex(i,j)){
					count++;
				}
				else if(arr2D[i][j]!=rightSide.getIndex(i,j)){
					return false;
				}	
			}
		}
		if(count==row*column) 
			return true;	
	}

	/*() operator overloading, returns the number in given indexes if they're valid*/
	int BoardArray2D::operator ()(int index1,int index2){
		
		if((index1 >= 0 && index1 <= getRow()) && (index2>=0 && index2<=getColumn()))
			return arr2D[index1][index2];
		else 
			return -1;
	}

	//Returns the coordinates of empty cell
	void BoardArray2D::getEmptyCellX_Y(int &cellX,int &cellY){
		for(auto i=0; i<row;i++){
			for(auto j=0;j<column;j++){
				if(arr2D[i][j]==100){
					cellX=i;
					cellY=j;
				}
			}
		}
	}
	
}//end of namespace

File: HW7/BoardArray2D.java
import java.io.*;
import java.security.SecureRandom;
import java.util.Scanner;

/**
 * Class BoardArray2D is a subclass of AbstractBoard class
 * Implements all methods of super class
 * Keeps board configuration at 2D array
 */
public class BoardArray2D extends AbstractBoard {

    private int[][] board2d;
    private int row;
    private int column;

    public void setRow(int n_row){
        this.row=n_row;
    }
    public void setColumn(int n_column){
        this.column=n_column;
    }
    public int getRow(){
        return this.row;
    }
    public int getColumn(){
        return this.column;
    }


    /**
     * No parameter constructor
     */
    public BoardArray2D(){
        this.row=9;
        this.column=9;
        board2d = new int[row][column];
        for(int i=0;i<row;i++){
            for(int j=0;j<column;j++) {
                board2d[i][j]=-1;
            }
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * Constructor that takes size information
     * @param size for square shape board
     */
    public BoardArray2D(int size){
        this.row=size;
        this.column=size;
        board2d = new int[size][size];
        for(int i=0;i<row;i++){
            for(int j=0;j<column;j++) {
                board2d[i][j]=-1;
            }
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * Constructor that takes two parameters
     * @param n_row as row size
     * @param n_column as column size
     */
    public BoardArray2D(int n_row,int n_column){
        this.row=n_row;
        this.column=n_column;
        board2d = new int[n_row][n_column];
        for(int i=0;i<row;i++){
            for(int j=0;j<column;j++) {
                board2d[i][j]=-1;
            }
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * toString method to print board configuration
     * @return String board configuration
     */
    @Override
    public String toString() {
        String result="";
        for (int m = 0; m < getRow(); m++){
            for (int k = 0; k < getColumn(); k++) {
                if(board2d[m][k]!=100)
                    result+= board2d[m][k] +" ";
                else if(board2d[m][k]==100)
                    result+= "bb ";
            }
            result+="\n";
        }
        return result;
    }

    /**
     * readFromFile reads file ans assigns the values to board array
     * @param filename as name of the file
     */
    @Override
    void readFromFile(String filename) {
        String[] readen =new String[100];
        try {
            int iRow=0;
            int jCol=0;
            int indexReaden=0;
            Scanner sc = new Scanner(new BufferedReader(new FileReader("puzzle.txt")));
            //Reading line by line and parsing those lines
            while (sc.hasNextLine()) {
                String[] line = sc.nextLine().split(" ");
                for(int i=0;i<line.length;i++) {
                    jCol=line.length;
                    line[i] = line[i].trim();
                    readen[indexReaden]=line[i];
                    indexReaden++;
                }
                iRow++;
            }
            //Updating board arrays size according to the file
            setSize(iRow,jCol);
            //Assigning readen values to board array
            for(int i=0;i<getRow();i++){
                for(int j=0;j<getColumn();j++){
                    try {
                        //Converting string to integer
                        if (readen[i * getColumn() + j] != "bb")
                            board2d[i][j] = Integer.parseInt(readen[i*getColumn() + j]);
                    }
                    catch (NumberFormatException nfe){
                        board2d[i][j] = 100;
                    }
                }
            }
        }
        catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    /**
     * writeToFile method to load board configuration
     * @param filename as name of the file
     */
    @Override
    void writeToFile(String filename) {
        try {
            FileOutputStream output = new FileOutputStream("output.txt");
            output.write(toString().getBytes());
            output.close();
        } catch (IOException e) {
            System.out.println("Error - " + e.toString());
        }
    }

    /**
     * reset method,resets board to the solution
     */
    @Override
    void reset() {
        int num=1;
        for(int i=0;i<getRow();i++){
            for(int j=0; j<getColumn();j++){

                if(i==getRow()-1 && j==getColumn()-1 && board2d[i][j]!=0){
                    board2d[i][j]=100;
                }
                else if( board2d[i][j]!=0){
                    board2d[i][j]=num;
                    num++;
                }
            }
        }
    }
    /**
     * setSize sets size,and resets the board
     * @param size1 as row
     * @param size2 as column
     */
    @Override
    void setSize(int size1, int size2) {
        setRow(size1);
        setColumn(size2);
        board2d = new int[size1][size2];
        for(int i=0;i<getRow();i++){
            for(int j=0;j<getColumn();j++){
                board2d[i][j]=-1;
            }
        }
        reset();
    }

    /**
     * move empty cell if there is a place
     * @param direction U R L D directions
     * @return 1 if move is done,0 if no place
     */
    @Override
    int move(char direction) {
        int emptCellX=-2;
        int emptCellY=-2;
        int num;

        for(int i=0; i<row; i++){
            for(int j=0; j<column; j++){
                if(board2d[i][j]==100){
                    emptCellX=i;
                    emptCellY=j;
                }
            }
        }
        if (direction=='U'){

            if(emptCellX==0 || board2d[emptCellX-1][emptCellY]==0){
                return -1;
            }
            else if(emptCellX!=0 && board2d[emptCellX-1][emptCellY]!=0){
                num=board2d[emptCellX-1][emptCellY];
                board2d[emptCellX-1][emptCellY]=100;
                board2d[emptCellX][emptCellY]=num;
                setLastMove('U');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction== 'D'){

            if(emptCellX==row-1 || board2d[emptCellX+1][emptCellY]==0){
                return -1;
            }
            else if(emptCellX!=row-1 && board2d[emptCellX+1][emptCellY]!=0){
                num=board2d[emptCellX+1][emptCellY];
                board2d[emptCellX+1][emptCellY]=100;
                board2d[emptCellX][emptCellY]=num;
                setLastMove('D');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction=='R'){

            if(emptCellY==column-1 || board2d[emptCellX][emptCellY+1]==0){
                return -1;
            }

            else if(emptCellY!=column-1 && board2d[emptCellX][emptCellY+1]!=0){
                num=board2d[emptCellX][emptCellY+1];
                board2d[emptCellX][emptCellY+1]=100;
                board2d[emptCellX][emptCellY]=num;
                setLastMove('R');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction=='L'){

            if(emptCellY==0 || board2d[emptCellX][emptCellY-1]==0){
                return -1;
            }

            else if(emptCellY!=0 && board2d[emptCellX][emptCellY-1]!=0){
                num=board2d[emptCellX][emptCellY-1];
                board2d[emptCellX][emptCellY-1]=100;
                board2d[emptCellX][emptCellY]=num;
                setLastMove('L');
                setNumOfMove();
            }
            return 0;
        }
        return 0;
    }

    /**
     * shuffle method,written to shuffle the board
     * @return int 0 if shuffle is done
     */
    public int shuffle(){

        /*First reset the board to the solution*/
        reset();
        //Initialize the random object
        SecureRandom random = new SecureRandom();

        int randomMove;
        int control=0;
        int i=0;
		/*This loop written for shuffle the board making ramdom moves size*size.
		But if there is an impossible move, i call random func. again to make sure
		the board is shuffled properly */
        while( i<getRow()*getColumn()){

            randomMove = random.nextInt(4) + 1;
            if(randomMove==1){
                control=move('U');
            }
            if(randomMove==2){
                control=move('D');
            }
            if(randomMove==3){
                control=move('L');
            }
            if(randomMove==4){
                control=move('R');
            }
		/*Control is the return value for U,D,R,L functions
		If control is -1 this means there is no room for that choice
		If control is 0 this means cell is moved and size*size number increases*/
            if(control==0)
                i++;
        }
        return 0;
    }

    /**
     * isSolved method to check if board is solution
     * @return true if board is solution,else false
     */
    @Override
    boolean isSolved() {
        BoardArray2D controlObje= new BoardArray2D (row,column);
        controlObje.reset();

        int count=0;
        for(int i=0; i<row; i++){
            for(int j=0; j<column; j++){
                if(board2d[i][j]==controlObje.cell(i,j)){
                    count++;
                }
                else if(board2d[i][j]!=controlObje.cell(i,j)){
                    System.out.println("Problem is not solved.\n");
                    return false;
                }
            }
        }
        if(count==row*column) {
            System.out.println("Problem is solved!\n");
            return true;
        }
        else
            return false;
    }

    /**
     * method cell gives the index in given coordinates
     * @param index1 as row
     * @param index2 as column
     * @return int cell content at given indexes
     */
    @Override
    int cell(int index1, int index2) {
        if(index1<0 || index1>this.row || index2<0 || index2>this.column) {
            System.out.println("Invalid index values!!");
            System.exit(0);
        }
        return board2d[index1][index2];
    }
    /**
     * Abstract method equals,if two objects are the same,boards are the same
     * @param other AbstractBoard object
     * @return true if objects are same,else return false
     */
    @Override
    boolean equals(AbstractBoard other) {
        // If the object is compared with itself then return true
        if (other == this) {
            return true;
        }
        //Check if other is an instance of BoardArray2D or not
        if (!(other instanceof BoardArray2D)) {
            return false;
        }
       // BoardArray2D control= new BoardArray2D(this.getRow(),this.getColumn());
        // typecast other so that we can compare data members
        BoardArray2D control = (BoardArray2D) other;
        if ((control.getColumn()!=getColumn()) || (control.getRow()!=getRow()))
            return false;
        if((control.getColumn()==getColumn()) && (control.getRow()==getRow()) ){
            for(int i=0;i<getRow();i++){
                for(int j=0;j<getColumn();j++){
                    if(board2d[i][j]!=control.cell(i,j))
                        return false;
                }
            }
            return true;
        }
        return false;
    }

    /**
     * getEmptyCellX method for using in global test function
     * @return int emptycell's x index
     */
    @Override
    public int getEmptyCellX(){
        for(int i=0; i<getRow();i++){
            for(int j=0;j<getColumn();j++){
                if(board2d[i][j]==100)
                    return i;
            }
        }
        return -1;
    }

    /**
     * getEmptyCellY method for using in global test function
     * @return int emptycell's y index
     */
    @Override
    //Returns the coordinates of empty cell
    public int getEmptyCellY(){
        for(int i=0; i<getRow();i++){
            for(int j=0;j<getColumn();j++){
                if(board2d[i][j]==100)
                    return j;
            }
        }
        return -1;
    }
}//end of the class BoardArray2D
